src/Servers/Kestrel/Core/src/KestrelServer.cs

```diff
@@ -211,73 +211,135 @@ public async Task StopAsync(CancellationToken cancellationToken)
-                    tasks[i] = Task.WhenAll(listener.UnbindAsync(cancellationToken).AsTask(), acceptLoop);
+                    reloadToken = Options.ConfigurationLoader.Configuration.GetReloadToken();
                 }
 
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+                Options.ConfigurationLoader?.Load();
 
-                if (!await ConnectionManager.CloseAllConnectionsAsync(cancellationToken).ConfigureAwait(false))
-                {
-                    Trace.NotAllConnectionsClosedGracefully();
+                await AddressBinder.BindAsync(Options.ListenOptions, AddressBindContext).ConfigureAwait(false);
+                _configChangedRegistration = reloadToken?.RegisterChangeCallback(async state => await ((KestrelServer)state).RebindAsync(), this);
+            }
+            finally
+            {
+                _bindSemaphore.Release();
+            }
+        }
 
-                    if (!await ConnectionManager.AbortAllConnectionsAsync().ConfigureAwait(false))
-                    {
-                        Trace.NotAllConnectionsAborted();
-                    }
-                }
+        private async Task RebindAsync()
+        {
+            await _bindSemaphore.WaitAsync();
+
+            IChangeToken reloadToken = null;
 
-                for (int i = 0; i < connectionTransportCount; i++)
+            try
+            {
+                if (_stopping == 1)
                 {
-                    (IConnectionListener listener, Task acceptLoop) = _transports[i];
-                    tasks[i] = listener.DisposeAsync().AsTask();
+                    return;
                 }
 
-                for (int i = connectionTransportCount; i < totalTransportCount; i++)
+                reloadToken = Options.ConfigurationLoader.Configuration.GetReloadToken();
+                var (endpointsToStop, endpointsToStart) = Options.ConfigurationLoader.Reload();
+
+                if (endpointsToStop.Count > 0)
                 {
-                    (IMultiplexedConnectionListener listener, Task acceptLoop) = _multiplexedTransports[i - connectionTransportCount];
-                    tasks[i] = listener.DisposeAsync().AsTask();
+                    var urlsToStop = endpointsToStop.Select(lo => lo.EndpointConfig.Url ?? "<unknown>");
+                    Trace.LogInformation("Config changed. Stopping the following endpoints: {endpoints}", string.Join(", ", urlsToStop));
+
+                    // 5 is the default value for WebHost's "shutdownTimeoutSeconds", so use that.
+                    using var timeoutCts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+                    using var combinedCts = CancellationTokenSource.CreateLinkedTokenSource(_stopCts.Token, timeoutCts.Token);
+
+                    // TODO: It would be nice to start binding to new endpoints immediately and reconfigured endpoints as soon
```

Comment:
Except there may be conflicts.

Reply:
Ofc. We would have to write the algorithm to sequence unbinding and binding only for endpoints that are being reconfigured instead of newly added. If it was a simple as binding and unbinding in parallel. I would have already done it.

That's probably not worth the effort. What might still be worth doing is adding a TransportManager method that just unbinds endpoints. We could unbind first, then bind in parallel to draining the old connections.

Reply:
We're going to get asked if non-destructive updates are possible. E.g. if all I did was change the cert file, why do I have to drop the socket? This will make even more sense when we get to config based SNI, if all I do is add one more host/cert to the list why should everything be torn down.

Not saying this is something we have to address right now, but we should at least think about if all changes are equal.